File: app/src/main/java/com/doapps/habits/helper/HomeDayManager.java
package com.doapps.habits.helper;

import com.doapps.habits.models.Habit;
import com.doapps.habits.models.IDayManager;
import com.doapps.habits.models.IUpdatableList;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class HomeDayManager implements IDayManager, DueCounter {
    private final IUpdatableList<Habit> mTimeLineAdapter;
    private final List<Habit> mHabitList;
    private int dueCount;

    public HomeDayManager(IUpdatableList<Habit> timeLineAdapter, List<Habit> habitList) {
        mTimeLineAdapter = timeLineAdapter;
        mHabitList = habitList;
    }

    @Override
    public void updateListByDayOfWeek(int dayOfWeek) {
        List<Habit> dayHabits = new ArrayList<>(mHabitList);
        DayFilter.filterListByDay(dayHabits, dayOfWeek);
        dueCount = countDue(dayHabits);
        mTimeLineAdapter.updateList(dayHabits);
    }

    @Override
    public void updateForToday() {
        List<Habit> todayHabits = new ArrayList<>(mHabitList);
        DayFilter.filterListForToday(todayHabits);
        dueCount = countDue(todayHabits);
        mTimeLineAdapter.updateList(todayHabits);
    }

    private int countDue(List<Habit> list){
        Calendar calendar = Calendar.getInstance();
        int date = calendar.get(Calendar.DATE);
        int month = calendar.get(Calendar.MONTH);
        int year = calendar.get(Calendar.YEAR);

        int counter = 0;
        for (Habit habit : list) {
            if (!habit.isDone(date, month, year)) {
                counter++;
            }
        }
        return counter;
    }

    @Override
    public String getDueCount(){
        return String.valueOf(dueCount);
    }


}

File: app/src/main/java/com/doapps/habits/helper/HomeDayManager.kt
package com.doapps.habits.helper

import com.doapps.habits.models.Habit
import com.doapps.habits.models.IDayManager
import com.doapps.habits.models.IListUpdater
import java.util.*

class HomeDayManager(private val mTimeLineAdapter: IListUpdater<Habit>,
                     private val mHabitList: List<Habit>) : IDayManager, DueCounter {
  private var dueCount: Int = 0

  override fun updateListByDayOfWeek(dayOfWeek: Int) {
    val dayHabits = ArrayList(mHabitList)
    DayFilter.filterListByDay(dayHabits, dayOfWeek)
    dueCount = countDue(dayHabits)
    mTimeLineAdapter.updateList(dayHabits)
  }

  override fun updateForToday() {
    val todayHabits = ArrayList(mHabitList)
    DayFilter.filterListForToday(todayHabits)
    dueCount = countDue(todayHabits)
    mTimeLineAdapter.updateList(todayHabits)
  }

  private fun countDue(list: List<Habit>): Int {
    val calendar = Calendar.getInstance()
    val date = calendar.get(Calendar.DATE)
    val month = calendar.get(Calendar.MONTH)
    val year = calendar.get(Calendar.YEAR)
    return list.count { !it.isDone(date, month, year) }
  }

  override fun getDueCount() = dueCount.toString()
}